### PR TITLE
Implement Nested Structure counting for C++

### PR DESCRIPTION
**Pull request by rakhimov**
The description of the metric is in Issue #69 and in the documentation.

---
This change is [Reviewable](https://reviewable.io/reviews/terryyin/lizard/119)

### CODE REVIEW

**Comment by mehrdad89** [comment]
Great addition! let's see what @terryyin think about the changes!

File: test/testNestedStructures.py [hunk 1]
```diff
@@ -0,0 +1,211 @@
...
+        result = process_cpp("""
+        x c() {
+          {{{{{{for (a : c) {
+            {{{{if(a != 0) {{{{
+                a = b;
+            }}}}}}}}
+          }}}}}}}
+        }
+        """)
+        self.assertEqual(2, result[0].max_nested_structures)
```
Comment (terryyin) [inline test/testNestedStructures.py]: Why the nesting with 'bare brackets' are ignored? There usually need to be good reason for having them. And in C++, they actually trigger destructors.

**Comment by terryyin** [comment]
great! @rakhimov , I like it very much.

So far this implementation works only for "C-like" languages. I have implemented some more general data for nesting.
At any time, you can get the current nesting level via `context. self.current_nesting_level` and
`context.current_function.top_nesting_level` keeps the top nesting level of the current function. Perhaps you can refactor the code to use these two values to calculate the NS result. Then it will become a general implementation for all supported languages.

I like the theory very much. I or somebody else should add CCN to it.

File: lizard_ext/lizardns.py [hunk 2]
```diff
@@ -0,0 +1,137 @@
...
+        """
+        structures = set(['if', 'else', 'foreach', 'for', 'while', 'do',
+                          'try', 'catch', 'switch'])
+
+        structure_indicator = "{"
+        structure_end = "}"
+        indent_indicator = ";"
+
+        for token in tokens:
+            if reader.context.is_within_structure():
```
Comment (terryyin) [inline lizard_ext/lizardns.py]: If you do everything in this one method, there's no need to add information to `reader.context`. A local variable/object in this method might serve the purpose already, if the state is not going to be shared with other processors.

Comment (rakhimov) [inline lizard_ext/lizardns.py]: That indeed simplifies the code.
By the way, is it possible to get brace/bracket counts from the reader?

**Comment by rakhimov** [comment]
Hi @terryyin ,
you have merged this PR too early.
It is quiet annoying to submit patches to patches.
I would commit additional patches within this PR to get your reviews into code,
which makes it more coherent.
I wish GitHub had 'unmerge' feature.

File: test/testNestedStructures.py (hunk 1 above)
Comment (rakhimov) [inline test/testNestedStructures.py]: Technically it is not control-flow structure,
but I think it is a good idea to include bare scope into this metric.
However, the bare braces have more to it than control-flow structures.
For example, wrapping your code with openmp pragma to get it parallelized when built with openmp; the code that does the job stays the same for all builds.

**Comment by rakhimov** [comment]
``context.current_nesting_level`` and ``current_function.top_nesting_level`` are undocumented.
It is not clear what ``nesting level`` means.
For example, the ``current_nesting_level`` produce different nesting for the following code:

```cpp
if (x) while(y) continue;

if (x) { while(y) { continue } }
```
In the first line of code, all ``if``, ``while``, ``continue`` get nesting 1.
In the second line of code, the nesting output is: (``if``, 1), (``while``, 2), (``continue``, 3).
Is it just brace counting for C++?

**Pull request merged**